src/storage/compressed_tile.cpp

```diff
@@ -0,0 +1,588 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// compressed_tile.cpp
+//
+// Identification: src/storage/compressed_tile.cpp
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <cstdio>
+#include <sstream>
+
+#include "catalog/schema.h"
+#include "common/exception.h"
+#include "common/macros.h"
+#include "type/serializer.h"
+#include "type/types.h"
+#include "type/ephemeral_pool.h"
+#include "type/value_factory.h"
+#include "concurrency/transaction_manager_factory.h"
+#include "storage/storage_manager.h"
+#include "storage/tile.h"
+#include "storage/tile_group_header.h"
+#include "storage/tuple.h"
+#include "storage/tuple_iterator.h"
+#include "storage/compressed_tile.h"
+#include "container/cuckoo_map.h"
+
+namespace peloton {
+namespace storage {
+
+bool CompareLessThanBool(type::Value left, type::Value right) {
+  return left.CompareLessThan(right);
+}
+
+// Constructor: Call base constructor and set is_compressed to false.
+CompressedTile::CompressedTile(BackendType backend_type,
+                               TileGroupHeader *tile_header,
+                               const catalog::Schema &tuple_schema,
+                               TileGroup *tile_group, int tuple_count)
+    : Tile(backend_type, tile_header, tuple_schema, tile_group, tuple_count) {
+  is_compressed = false;  // tile is currently uncompressed.
+  compressed_columns_count = 0;
+}
+
+CompressedTile::~CompressedTile() {}
+
+// CompressTile
+// @brief: This function identifies the columns that can be compressed and
+//         determines the smallest compression type that can be used.
+//         The compression scheme for each type:
+//           1. INTEGER Types other than TINYINT - Find the median and store
+//              it as the base_value, store all the remaining values in types
+//              of smaller sizes.
+//           2. DECIMAL - Same as INTEGER Types, additionally store the
+//              maximum exponent.
+//           3. VARCHAR - Deduplicate the VARCHAR columns using a Cuckoo Hash.
+//        For each compressed column, store the metadata for retrieving the
+//        original values.
+
+void CompressedTile::CompressTile(Tile *tile) {
```

Comment:
This function is too long. I would break it down and add some private helper function to make it more readable.